File: cluster_PETs.cpp
#define BOOST_SPIRIT_DEBUG

#include <fstream>
#include <cstdio>

#include <boost/fusion/adapted/struct.hpp>
#include <boost/spirit/include/qi.hpp>
#include <boost/iostreams/device/mapped_file.hpp>
#include <boost/utility/string_ref.hpp>
#include <boost/program_options.hpp>

namespace qi = boost::spirit::qi;
using sref   = boost::string_ref;
namespace po = boost::program_options;
using namespace std;

namespace boost { namespace spirit { namespace traits {
    template <typename It>
    struct assign_to_attribute_from_iterators<sref, It, void> {
        static void call(It f, It l, sref& attr) { attr = { f, size_t(std::distance(f,l)) }; }
    };
} } }

struct PET {
    sref chrom1;
    int start1, end1;
    sref chrom2;
    int start2, end2;
    int count;

    bool operator<(const PET& a) const
    {
        if (chrom1 != a.chrom1) return chrom1 < a.chrom1;
        if (start1 != a.start1) return start1 < a.start1;
        if (end1 != a.end1) return end1 < a.end1;
        if (chrom2 != a.chrom2) return chrom2 < a.chrom2;
        if (start2 != a.start2) return start2 < a.start2;
        if (end2 != a.end2) return end2 < a.end2;
        return false;
    }

    friend ostream& operator<<(ostream& os, const PET& pet)
    {
        os << pet.chrom1 << ":" << pet.start1 << "-" << pet.end1 << " " << pet.chrom2 << ":" << pet.start2 << "-" << pet.end2 << " (" << pet.count << ")";
        return os;
    }    
};

BOOST_FUSION_ADAPT_STRUCT(PET, (sref,chrom1)(int,start1)(int,end1)(sref,chrom2)(int,start2)(int,end2)(int,count))

int main(int ac, char** av) {
    string input_filename, output_filename;
    int pet_cutoff, cluster_cutoff, extension, self_ligation;

    try
    {
        po::options_description desc("Allowed options");
        desc.add_options()
            ("help,h", "help message")
            ("input,i", po::value<string>(&input_filename)->required(), ".bedpe file containing raw PETs (no header)")
            ("output,o", po::value<string>(&output_filename)->required(), ".bedpe output file with clustered PETs (no header)")
            ("self_ligation,s", po::value<int>(&self_ligation)->default_value(8000), "Self-ligation genomic span (default: 8000)")
            ("extension,e", po::value<int>(&extension)->default_value(500), "No of base pairs to extend both ends of PETs (default: 500)")
            ("pet_cutoff,p", po::value<int>(&pet_cutoff)->default_value(2), "Minimum number of PET counts to take PET into consideration (default: 2)")
            ("cluster_cutoff,c", po::value<int>(&cluster_cutoff)->default_value(4), "Minimum number of total counts to consider as a cluster (default: 4)");

        po::variables_map vm;
        po::store(po::parse_command_line(ac, av, desc), vm);
        if (vm.count("help")) {
            cout << desc << "\n";
            return 0;
        }
        po::notify(vm);
    }catch (const po::error &ex)
    {
        cerr << ex.what() << '\n';
        return 0;
    }

    // reading the file
    cout << "Reading PETs from " << input_filename << endl;
    boost::iostreams::mapped_file_source mmap(input_filename);

    // parsing
    using namespace qi;
    vector<PET> pets;
    pets.reserve(44455464);
    if (parse(mmap.begin(), mmap.end(), 
                (raw[*~char_("\t")] >> '\t' >> int_ >> '\t' >> int_ >> '\t' >> raw[*~char_("\t")] >> '\t' >> int_ >> '\t' >> int_ >> '\t' >> int_) % eol, 
                pets))
    {
        cout << "Parsed " << pets.size() << " PETs" << endl;
    } else {
        cout << "Failed after " << pets.size() << " PETs" << endl;
    };

    // sorting
    cout << "Sorting... " << flush;
    sort(pets.begin(), pets.end());
    cout << "Done." << endl;

    // pre-proccess
    cout << "Preprocessing (Extension: "<< extension << "bp, Self-ligation genomic span: " << self_ligation << "bp, PET cutoff: " << pet_cutoff << ")... " << flush;
    for (int i = 0; i<pets.size(); i++){
        PET &pet = pets[i];
        // check the data integrity
        if (pet.chrom1 != pet.chrom2 || pet.start1 > pet.end1 || pet.start2 > pet.end2){
            cout << "Inter-chromosomal or misordered PETs are ignored: #" << (i+1) << " " << pet << endl;
            pet.count = 0;
        }

        // remove self ligating PETs
        if (pet.start2 - pet.end1 < self_ligation)
            pet.count = 0;

        // remove PETs with count < PET cutoff
        if (pet.count < pet_cutoff)
            pet.count = 0;

        // add extension
        pet.start1 = max(0, pet.start1 - extension);
        pet.end1 = pet.end1 + extension;
        pet.start2 = max(0, pet.start2 - extension);
        pet.end2 = pet.end2 + extension;
    }
    cout << "Done." << endl;

    // clustering
    cout << "Clustering... " << flush;
    for (int i =0; i<pets.size(); i++) {
        if (pets[i].count == 0) 
            continue;
        int j = i + 1;
        PET &cluster = pets[i];
        while (j < pets.size()) {
            PET &can = pets[j];
            if (can.start1 > cluster.end1 or can.chrom1 != cluster.chrom1)
                break;
            if
            ((((pets[i].start1 <= pets[j].start1) && (pets[j].start1 <= pets[i].end1)) || 
                 ((pets[i].start1 <= pets[j].end1) && (pets[j].end1 <= pets[i].end1))) &&
                (((pets[i].start2 <= pets[j].start2) && (pets[j].start2 <= pets[i].end2)) || 
                 ((pets[i].start2 <= pets[j].end2) && (pets[j].end2 <= pets[i].end2)))) 
                 {
                cluster.start1 = min(cluster.start1, can.start1);
                cluster.end1 = max(cluster.end1, can.end1);
                cluster.start2 = min(cluster.start2, can.start2);
                cluster.end2 = max(cluster.end2, can.end2);
                cluster.count += can.count;
                can.count = 0;
            }
            j++;            
        }
    }
    cout << "Done." << endl;

    // save to file
    cout << "Saving to " << output_filename << " (cluster cufoff: " << cluster_cutoff << ")... " << flush;
    ofstream ofs(output_filename, ofstream::out);
    char buff[64];
    int c = 0;
    for (int i =0; i<pets.size(); i++){
        PET &cluster = pets[i];
        if (cluster.count >= cluster_cutoff){
            std::string chrom1(cluster.chrom1.data(), cluster.chrom1.size()), chrom2(cluster.chrom2.data(), cluster.chrom2.size());
            sprintf(buff, "%s\t%d\t%d\t%s\t%d\t%d\t%d\n", chrom1.c_str(), cluster.start1, cluster.end1, chrom2.c_str(), cluster.start2, cluster.end2, cluster.count);
            ofs << buff;
            c++;
        }
    }
    ofs.close();
    cout << "Done. Saved " << c << " clusters." << endl;
}

File: cluster_pets/cluster_PETs.py
import argparse
import logging
import pandas as pd
import numpy as np
import numba
import itertools
from pyranges import PyRanges

logging.basicConfig(level=logging.INFO, format='%(asctime)s %(message)s', datefmt='%H:%M:%S')


def add_build_args(parser):
    parser.add_argument("--pets_filename", type=str, action='append',
                        default=[],
                        help=".bedpe files containing raw PETs (no header)")
    parser.add_argument("--clusters_filename", type=str,
                        default="~/BioData/chromatin_loops/4DNFI2BAXOSW_GM12878_CTCF_rep1_hiseq.bedpe.25.2.6.clusters",
                        help=".bedpe output file with clustered PETs (no header)")
    parser.add_argument("--peaks_filename", type=str,
                        help="Peaks which must intersect with PETs to be considered during clustering (.bed file)")
    parser.add_argument("--self_ligation", type=int, default=8000, help="Self-ligation genomic span (default: 8000)")
    parser.add_argument("--extension", type=int, default=25,
                        help="No of base pairs to extend both ends of PETs (default: 25)")
    parser.add_argument("--pet_cutoff", type=int, default=2,
                        help="Minimum number of PET counts to take PET into consideration (default: 2)")
    parser.add_argument("--cluster_cutoff", type=int, default=6,
                        help="Minimum number of total counts to consider as a cluster (default: 6)")
    parser.add_argument("--nrows", type=int,
                        help="If provided limits the number of rows read from the PET file to nrows")


def cluster_PETs(args):
    # reading the file
    logging.info(f"Reading PETs from {args.pets_filename} ...")
    columns = ["chrom1", "start1", "end1", "chrom2", "start2", "end2", "cnt"]
    pets = pd.concat([pd.read_csv(f, sep="\t", header=None, names=columns, low_memory=False, nrows=args.nrows) for f in args.pets_filename])
    logging.info(f"Read {len(pets):,} PETs.")

    # pre-proccess
    logging.info(f"Preprocessing (Exstension: {args.extension}bp, Self-ligation genomic span: {args.self_ligation}bp, "
                 f"PET cutoff: {args.pet_cutoff}) ...")

    # check the data integrity
    invalid_pets = pets[(pets.chrom1 != pets.chrom2) | (pets.start1 > pets.end1) | (pets.start2 > pets.end2)]
    if len(invalid_pets) > 0:
        logging.info(f"{len(invalid_pets)} inter-chromosomal or misordered PETs are ignored:")
        logging.info(invalid_pets.head())
    pets = pets[(pets.chrom1 == pets.chrom2) & (pets.start1 <= pets.end1) & (pets.start2 <= pets.end2)]

    # keep only non-self-ligating PETs
    pets = pets[pets.start2 - pets.end1 >= args.self_ligation]

    # keep PETs with count >= PET cutoff
    pets = pets[pets.cnt >= args.pet_cutoff]

    # add extension
    pets.start1 = (pets.start1 - args.extension).clip(0)
    pets.end1 = pets.end1 + args.extension
    pets.start2 = (pets.start2 - args.extension).clip(0)
    pets.end2 = pets.end2 + args.extension

    # remove not intersecting anchors
    if args.peaks_filename:
        peaks = pd.read_csv(args.peaks_filename, sep="\t", header=None, usecols=[0, 1, 2, 6],
                            names=["Chromosome", "Start", "End", "Score"])
        peaks = PyRanges(peaks)
        pets["Chromosome"], pets["Start"], pets["End"] = pets.chrom1, pets.start1, pets.end1
        pets = PyRanges(pets).intersect(peaks).df
        pets["Chromosome"], pets["Start"], pets["End"] = pets.chrom2, pets.start2, pets.end2
        pets = PyRanges(pets).intersect(peaks).df
        peaks = peaks.df
        peaks["Center"] = peaks["Start"] + (peaks["End"]-peaks["Start"]) // 2

    logging.info(f"Done. {len(pets):,} PETs left.")

    chroms = pets.groupby(["chrom1"]).size().to_dict()

    start1s = numba.typed.List([pets[pets.chrom1 == chrom].start1.to_numpy() for chrom in chroms.keys()])
    end1s = numba.typed.List([pets[pets.chrom1 == chrom].end1.to_numpy() for chrom in chroms.keys()])
    start2s = numba.typed.List([pets[pets.chrom1 == chrom].start2.to_numpy() for chrom in chroms.keys()])
    end2s = numba.typed.List([pets[pets.chrom1 == chrom].end2.to_numpy() for chrom in chroms.keys()])
    cnts = numba.typed.List([pets[pets.chrom1 == chrom].cnt.to_numpy() for chrom in chroms.keys()])

    step = 0
    changes = np.ones(shape=(len(chroms),))
    while np.sum(changes) > 0:
        # sorting# if changes[i] > 0 else None\
        logging.info(f"Sorting (step: #{step+1}, PETs: {len(pets):,}) ...")
        orders = numba.typed.List([np.lexsort((end2s[i], start2s[i], end1s[i], start1s[i]))
                                   for i in range(len(chroms))])
        logging.info("Done.")

        # clustering
        logging.info(f"Clustering (step: #{step+1}, PETs: {len(pets):,}) ...")

        @numba.jit(nopython=True, parallel=True)
        def cluster(chroms, old_changes, orders, start1s, end1s, start2s, end2s, cnts):
            changes = np.zeros(shape=(len(chroms)-1,), dtype=np.uint64)

            for idx in numba.prange(len(chroms)-1):
                if old_changes[idx] > 0:
                    order, start1, end1, start2, end2, cnt =\
                        orders[idx], start1s[idx], end1s[idx], start2s[idx], end2s[idx], cnts[idx]
                    for _i in range(chroms[idx]):
                        i = order[_i]
                        if cnt[i] == 0:
                            continue
                        _j = _i + 1
                        while _j < chroms[idx]:
                            j = order[_j]
                            if start1[j] > end1[i]:
                                break
                            if cnt[j] == 0:
                                _j += 1
                                continue
                            if ((start1[i] <= start1[j] and start1[j] <= end1[i]) or (start1[i] <= end1[j] and end1[j] <= end1[i])) and\
                               ((start2[i] <= start2[j] and start2[j] <= end2[i]) or (start2[i] <= end2[j] and end2[j] <= end2[i])):
                                start1[i] = min(start1[i], start1[j])
                                end1[i] = max(end1[i], end1[j])
                                start2[i] = min(start2[i], start2[j])
                                end2[i] = max(end2[i], end2[j])
                                cnt[i] += cnt[j]
                                cnt[j] = 0
                                changes[idx] += 1
                            _j += 1
            return changes
        changes = cluster(numba.typed.List(chroms.values()), changes, orders, start1s, end1s, start2s, end2s, cnts)
        logging.info(f"Done. Changes: {int(sum(changes)):,}")
        step += 1

    # save to file
    logging.info(f"Saving to {args.clusters_filename} (cluster cufoff: {args.cluster_cutoff})... ")
    pets = pd.DataFrame()
    for i, (chrom, size) in enumerate(chroms.items()):
        pets = pd.concat([
            pets, pd.DataFrame(data={"chrom1": itertools.repeat(chrom, size),
                                     "start1": start1s[i][orders[i]],
                                     "end1": end1s[i][orders[i]],
                                     "chrom2": itertools.repeat(chrom, size),
                                     "start2": start2s[i][orders[i]],
                                     "end2": end2s[i][orders[i]],
                                     "cnt": cnts[i][orders[i]]})])
    pets = pets[pets.cnt >= args.cluster_cutoff]
    if peaks:
        pets["Center1"] = pets.apply(lambda row: peaks.iloc[
                            peaks[(peaks.Chromosome == row.chrom1) &
                                  (((peaks.Start >= row.start1) & (peaks.Start < row.end1)) |
                                  ((peaks.End >= row.start1) & (peaks.End < row.end1)))]
                                 ["Score"].idxmax()]["Center"], axis=1)
        pets["Center2"] = pets.apply(lambda row: peaks.iloc[
                            peaks[(peaks.Chromosome == row.chrom1) &
                                  (((peaks.Start >= row.start2) & (peaks.Start < row.end2)) |
                                  ((peaks.End >= row.start2) & (peaks.End < row.end2)))]
                                 ["Score"].idxmax()]["Center"], axis=1)
    pets.to_csv(args.clusters_filename, sep="\t", index=False, header=False)
    logging.info(f"Done. Saved {len(pets):,} clusters.")

    return pets


def main():
    parser = argparse.ArgumentParser()
    add_build_args(parser)
    args = parser.parse_args()
    cluster_PETs(args)


if __name__ == "__main__":
    main()
